### ISSUE TITLE
It seems impossible to type HoCs with defaultProps after 0.57.x

### ISSUE THREAD

**Issue by rsolomon** [facebook/flow#5187]
After the React types were upgraded in 0.53, `defaultProps` were no longer being $Diff'd out of HoCs automatically, since they were no longer an explicit parameter of Components. Issue #4644 provided a reasonable custom solution.

After 0.57, The changes to `$Diff` have invalidated that approach and I cannot figure out a proper way to type an HoC, such that the output component:

1. Properly sets props provided in `defaultProps` as optional.
2. Works for both functional and class-based components.
3. Properly diffs out the props provided by the HoC in the output component.
4. Allows composition of HoCs.

[This is as close as I could get](https://flow.org/try/#0PTAEAEDMBsHsHcBQiCWBbADrATgF1AFSgCGAzqAEoCmxAxvpNrGqAOTY32sDcyVAHljyhcATwxVQAYWhlSU5lgB2VJbgA8ABSYZSAGlABlXMVxUAfKAC8iUNNmlS66nVwA6BZlgq1WnfqMTM3NzXkQBIXwxCXs5T2VVXAB1FFwACwARKkhiAFdoXG1YXXUsnPzC-wMi3QNjUwtrW1jHeO9Ev2KA+uDQADJQAG8AE2y8gprSAC5QMvHKroBfMIicKPFJNp9k1MyxisnS-Ymq0EnLGzsXeg9FdrUAFQ3O3UsBkeOF3Rm5g-9l5DhQRrEQbUAACRQAHM0lRsAB5bCjbBbDrUACOuRQHGGk2qTAAbihRrj-BdmgMABTqRHQlBKYjQPGzT7nSm0O7bGaotQpdK-E5dI7lQW1UC0qH0xnnACU1nMzTsV047h5uCeEmcVEx2KopK6-VAABJDGliJqMVicZNDRKpUyyYajRkUJBIDTsHSGQ6uvjYESSTaBgKvqQQjKKaBqXbvWyOV4uZQVbcE4kNVQPV7pWS5VZLNdVZy089Lbr9bonS63ZnJbHTkUA3rzuYZWFRrRZBxQATiNhQBhCcSqMZPUooQBGGaQmFwxHItXqQaLAyDUi4UcTmZrjeLUKIdudyQ9vsD-1DgByuTQACM4eOAIJKYaljiXm9wgBMU+hsIRSLhC6DEoV63tgX6gMB77YMuQyQaBk4QSBcK7rwB69kevagBwOqvkh2AAMzfjOf7zkWvhAXhhGIVBMFLnuSoMaAiAckoa6gAAaoyxIAGK5Eo9AoN4jJqjMBZuD0VDQFQji8fxuCCd6gHbvSm6gMpY4GHBd4zFp0EXFGp7fEM6mqSZml4QhumLLmljqMMKAEqAwD0UqiAgNI46gKowzkLkGCgGaRJjiIaQoOQ0RUDMblgAABmq6aLrpVFWYalIUVBll4eZUHJVlanripCEmYsoAAD5DKVumZVB2WgblNX5RuRUFWOpXWeYMXMd4bFSJ5VhYdqVpUG+dWUpx0A8XxAlCdAaqtoqjFdax+D8NYoDqL1jWFVYABEJk7dR8FWIM44lUlx34SVznIO5bh3aA8ChbQaSgGFoD2W6KC0BUoCQDgxpVpAIiwP2vakFQBikCDUgfl5T6+f5gUqVM0WgHFZHqs86WgeBKXvJVlE6XlJnNRubUdajACSKiCBNtDENe0CiL9OBQlQuDycF6SmIdn4kE+vMEV5-Bhe4S09bD-Wno2I13o+z6DbqstgZSvXzYxrksWxzP9RtsMk7t+2Cx+x2nYL+EXVde4LQxNuuRrdhuQQ4v4G04PDGqa3S0OI6FZS3ujMrD5Pi+w14R+lLYUNyv4WNXHDLJ02KRjMqp7wBDAEAA), but everything falls apart once composition comes into play. It also requires the explicit typing of stateless functional components.

Looking at the output types, it seems like:

1. It infers the generics from the instantiation of the component, not when the component is created from the HoC.
2. Creates a weird union of exact and inexact types, which contribute to the issues as the composed components attempt any further refinement.

**Comment by rsolomon** [comment]
This is about as close as I can get:

```
export type SomeComponent<DefaultProps, Props> = ComponentWithDefaultProps<DefaultProps, Props> | React.ComponentType<Props>;        
 
type HigherOrderComponent<RequiredProps, ProvidedProps> = 
  (<Props, DefaultProps>(component: SomeComponent<DefaultProps, ProvidedProps & Props>) => 
        React.ComponentType<RequiredProps & Props>);
```

I have access to `DefaultProps` this way, but any attempt to use them (via `$Diff` or otherwise) just seems to zap any type safety from the whole system.

**Comment by rsolomon** [comment]
I think we got a reasonable solution here ( https://github.com/digiaonline/react-flow-types/pull/16 ), if anyone is curious. While it would be nice if there were a built-in way to accomplish this, I'll close this to reduce noise.

**Comment by Blasz** [comment]
This doesn't seem to fix the issue for cases where defaultProps is a subtype of Props.

E.g.

```
type Props = {
  foo: string | number;
}

class myComp extends React.Component { ... }
myComp.defaultProps = {
  foo: ''
};
```

`$Diff` breaks since `Props` and `defaultProps` don't have the same types, see: [this example](https://flow.org/try/#0C4TwDgpgBACgTgezAZygXigbwFBSgMwQQC4plg4BLAOwHMoAfKagVwFsAjCOAbmwF9s2AMYJq5KABEI+AIYsANsHhJUGHHkIkoABgFDQkKZXz50UACSST+ADwqUAGiiGICM9LmLliFAD4+ETEJWVJrU3MNAiJSPX4eIA) for the Diff mismatch case and

[this repro case](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BudYMDCFSAaxoA7MM1ZgAFoUI5yALkEBzDITkBXAEYA6eq2AATDJtJwJMDBJrAuDQgFpYiF4QCeOGuXQ0AB54RMI+NGAAwqx49hKEAOpacgAiNFCkOjCEAAosygA8aRlZufmqYADeAL4ANGB5cMoqVdUAfGAAvKhg1NyEBtG41pKEACrhBY3KHQBkVabpmdnTFcXLZU3k1fyooCA94AASNFxgGJSkYOqSpxj0Yb7SBGBHAPKRqoeHYACSxCJxJRCAg4I8IjhSPhSFgaIRThVaABHHQYLimVZkCSmBosABuGEWGPK3xIyNR6NWLTGcgiUDspBgYAQ0JwvhxxmGcWICAwMCZ9hoOPJaKFq3qdjIpnMhAw1mEYK0lEQUhwJJIjQJRKpYBpdIZTJZpDZQrAnNio2ZfIFcGIgpxmsJYvK9T0NHomXIEW0EXekSt-LAargWu9tLYhDBPvOEns+DNMRG8UOaCCIQB4VeFlp+De+EW+CGFviBRFlPKLRq9Ud2orrQ63V68wAFAU8xYDeKwOtSqs2s3zUnCC0i0OktoeytykUlr2XWB25oJIy+wBKLptHq9be0ZyDRPcia+Us0FGi4lbMDzAAkKQwUCgBVv98fi8785rzq2bXqk82MyvMBrwAZTkY0aDbfAO2XGA+zaVctxbSDoJXcp+0HbkWl3Rh9y5UYjwgt8YLXDcwC3bc+j3UdD0mMsv2UQDnwfZCl1QrZq3xJ0LxmeDdlQbwnkxToqjIFoJB0LA3XjAAfMByEIKCJHUV0xIkqSwB2dB6BgChKAAQTAIJ4WxShsIGajRimNCqi3eTSFlB5FhKKdL2EypRLAABGDTdl6LhsVOZt10qciwC4QgdHwKQCnMPEwGATdemqVBkooijQvS7TdKiQzAmM0xTP6XDi0IKzvxsvzJALIKKu3cLIui2L4sSjSUoEIQci9HRTDgFw-WEMDiAAKx0eSg10vwBoiDDLW0FgdHUORUGMCQxuUGhurgI44HoFojmzU48wLCySyresujAZsTuHLEvFXFpSAkLxSJCyqIqiy7g2UddOg6AprqqAwga+7Zmv4ZLUAKAy9E6SpPIABmqMH4qEAA5MFThYfByHqChct8RghUhgzSFhhGkZhuHEeRwQwHRwz8CxnGyEoNN3XhUw9iEKGyE6AAichWAieTFPUPmwEp8maaEABRRmCEuVngnZonlusMaACELvWzbtvoZs9NXfhaeMRn2ZgZ7MYVyGtcl6mEt4FG6Yx+XsdxpWCY5m3eapimyfttpHdp+mrbdln8ZVznaYKLXSYFoW5IUuwxYl-2kYdp25aZ8O2cJzmMu3SH-VJ33U9LjPg5dpn3YjvOubAf6y6livZdd5m8dzjm0sL6Pi-5wXYUT0Xxbt9PA8ztuc+VvPC7V1biBSbWup6vWrqNgvekhxeS6l0fpedhnq6nz2iaKXm+agOA4BHtP96zhWa87oUgA) for the HOC type failing when using union types.

I've tried constraining the `DefaultProps` generic to `$Shape<Props>` in the `HigherOrderComponentType` but can't seem to get it to work. Probably because it's trying to reference a sibling generic type.

@rsolomon would you mind re-opening this? I don't believe there's an adequate solution for this.

**Comment by rsolomon** [comment]
@Blasz I wouldn't recommend using the HoC from `react-flow-types` at this point. It's quite brittle, and features added to Flow in the previous few versions have rendered it obsolete.

Using `React.ElementConfig<>` should get you where you need to go. [Here's the relevant bit in the documentation.](https://flow.org/en/docs/react/hoc/#toc-supporting-defaultprops-with-react-elementconfig)

In your example, the HoC would be written something like:

```
function pseudoHoc<Props, Component: React.ComponentType<Props>>(
  WrappedComponent: Component
): React.ComponentType<React.ElementConfig<Component>> {
  return props => <WrappedComponent {...props} />;
}
```

**Comment by Blasz** [comment]
@rsolomon I completely missed that part of the docs, thank you for that!

FYI for others that stumble across this, it seems like ElementConfig was introduced in 0.63 (there's no mention of it in the changelogs) but still had the bug with unions/subtypes that I was described here. That issue is fixed in 0.64 though and everything seems to work fine 🎉 .
